src/Migrations/HSLMigration.php

```diff
@@ -0,0 +1,646 @@
+      'sin' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'sin'),
+      'tan' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'tan'),
+      'sqrt' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'sqrt'),
+      'log' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'log'),
+      'min' => shape(
+        'ns' => HSL_NAMESPACE::Math,
+        'name' => 'min',
+        'has_overrides' => true,
+      ),
+      'max' => shape(
+        'ns' => HSL_NAMESPACE::Math,
+        'name' => 'max',
+        'has_overrides' => true,
+      ),
+
+      // Container functions
+      'count' => shape(
+        'ns' => HSL_NAMESPACE::C,
+        'name' => 'count',
+        'argument_order' => vec[0],
+      ),
+    ];
+
+
+  <<__Override>>
+  public function migrateFile(string $path, EditableNode $root): EditableNode {
+    // find all the function calls
+    $nodes = $root->getDescendantsOfType(FunctionCallExpression::class);
+
+    // keep track of any HSL namespaces we may have to add to the top of the file
+    $found_namespaces = dict[];
+
+    // check if any functions are in the replacement list
+    foreach ($nodes as $node) {
+
+      // bail if not in the config
+      $fn_name = $this->getFunctionName($node);
+      if (
+        $fn_name === null ||
+        !C\contains_key(self::PHP_HSL_REPLACEMENTS, $fn_name)
+      ) {
+        continue;
+      }
+
+      // found a function to replace!
+      $replace_config = self::PHP_HSL_REPLACEMENTS[$fn_name];
+      $namespace = $replace_config['ns'];
+      $replacement = $replace_config['name'];
+
+      // build the replacement AST node
+      $new_node =
+        $this->replaceFunctionName($node, $namespace.'\\'.$replacement);
+
+      // possibly change argument order
+      $argument_order = $replace_config['argument_order'] ?? null;
+      if (
+        $argument_order !== null || ($replace_config['has_overrides'] ?? false)
+      ) {
+        $new_node = $this->maybeMutateArguments(
+          $root,
+          $new_node,
+          $argument_order,
+          $path,
+          inout $found_namespaces,
+        );
+      }
+
+      // if we got null back here, it means the function call has unsupported arguments. forget it for now
+      if ($new_node === null) {
+        continue;
+      }
+
+      // we know we're rewriting the node, so now we know we need the namespace
+      $found_namespaces[$namespace] = true;
+
+      // replace it in the ast
+      $root = $root->replace($node, $new_node);
+
+      // potentially change adjacent expressions to check for null instead of false
+      if ($replace_config['replace_false_with_null'] ?? false) {
+        $root = $this->maybeChangeFalseToNull($root, $new_node);
+      }
+    }
+
+    if (C\count($found_namespaces) === 0) {
+      return $root;
+    }
+
+    // add "use namespace" declarations at the top if they aren't already present
+    $declarations =
+      vec($root->getDescendantsOfType(INamespaceUseDeclaration::class));
```

Comment:
This is ok for the common case, but:

```Hack
namespace Foo {
}

namespace Bar {
  use namespace HH\Lib\Vec;
}
```

... is also valid

Reply:
Should be addressed in https://github.com/hhvm/hhast/pull/108/commits/ba302ff77b6cea2ba4aae34fe7710048400071d9

I cleaned up the logic, using `insertBefore` to skip over namespace declarations, and also handled both kinds of namespace declarations (with and without body).

I didn't do anything special to handle two block namespace declarations in the same file, though. It will always insert it inside the first one.